citest/json_predicate/keyed_predicate_result.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implements a ValuePredicateResult containing multiple values by key."""
+
+from .predicate import (
+    CloneableWithNewSource,
+    PredicateResult)
+
+
+class KeyedPredicateResultBuilder(object):
+  """Helper class for building a keyed result."""
+
+  def __init__(self, pred):
+    self.__pred = pred
+    self.cause = None
+    self.comment = None
+    self.__results = {}
+
+  def add_result(self, key, result):
+    """Adds a result to the list of results captured.
+
+    Args:
+      key: [String] The key to add the result under should be unique.
+      result: [PredicateResult] The result to add.
+    """
+    if key in self.__results:
+      raise ValueError('{0} already exists.'.format(key))
+    self.__results[key] = result
+    return self
+
+  def update_results(self, results):
+    """Adds a dictionary of results to the composite list of results captured.
+
+    Args:
+      results: [dictionary of PredicateResult] The results to add.
+    """
+    self.__results.update(results)
+    return self
+
+  def build(self, valid):
+    """Construct the specified KeyedPredicateResult instance.
+
+    Args:
+      valid: [bool]  Whether the result is considered successful or not.
+    """
+    return KeyedPredicateResult(
+        valid, self.__pred, self.__results,
+        comment=self.comment, cause=self.cause)
+
+
+class KeyedPredicateResult(PredicateResult, CloneableWithNewSource):
+  """Aggregates a collection of predicate results into a single response.
+
+  Each of the individual results is accessible by a string key value.
+
+  Attributes:
+    pred: The ValuePredicate doing the aggregation.
+    results: The dictionary of PredicateResponse instances being aggregated.
+  """
+
+  @property
+  def pred(self):
+    """The predicate used to collect the composite results."""
+    return self.__pred
+
+  @property
+  def results(self):
+    """The map of PredicateResult instances."""
```

Comment:
At some point I think the object model of Citest should be mapped out in docs in a visual format. The model both quite nested and abstract, so it would benefit us (and other users of Citest) to have a reference to work from.